File: codes-and-outputs/analytical.py
import pandas as pd
import numpy as np

# LOAD THE DATA AND RENAME THE COLUMN
# ACCOUNT CSV FILE
df = pd.read_csv("../data/accounts.csv")

df.rename(columns={'id': 'account_id'}, inplace=True)
df.rename(columns={'date': 'open_date'}, inplace=True)

# REMOVE SPACES FROM COLUMN NAMES
df.rename(columns=lambda x: x.strip(), inplace=True)

# COUNT THE NUMBER OF MISSING VALUES IN EACH COLUMN; none
missing_values_num = df.isna().sum()

# TYPE CAST THE DATE COLUMN
df['open_date'] = pd.to_datetime(df['open_date'])

# LOAD DISTRCICT DATA AND GET THE DISTRICT NAME
district_df = pd.read_csv("./districts_py.csv")

df = pd.merge(df, district_df, on='district_id', how ='left')
df['district_id'] = df['name']

# CHANGE THE COLUMN NAME AND DROP THE COLUMNS
df.rename(columns={'district_id': 'district_name'}, inplace=True)
df = df.iloc[:, :4]

# LOAD THE LINKS DATA FOR NUM_CUSTOMERS COLUMN
link_df = pd.read_csv("../data/links.csv")
temp = pd.DataFrame(link_df.groupby('account_id')['client_id'].count())
temp = temp.rename(columns = {'client_id':'num_customers'})
transpose = temp.T
temp['account'] = transpose.columns

df= df.merge(temp, left_on = "account_id", right_on = "account", how = "left")
df = df.drop("account", axis = 1)

# CREDIT_CARDS
card_df = pd.read_csv("../data/cards.csv")
temp = link_df.merge(card_df, left_on = "id", right_on = "link_id", how = "left")

temp = pd.DataFrame(temp.groupby("account_id")['id_y'].count())
temp = temp.rename(columns = {"id_y": "credit_cards"})

column_extract = temp.T

column_extract
temp['account'] = column_extract.columns

pd.unique(temp['credit_cards'])

df = df.merge(temp, left_on = "account_id", right_on="account", how = "left")
df.drop("account", axis = 1, inplace = True)

# LOAN
loan_df = pd.read_csv("./loans_py.csv")
df = pd.merge(df,loan_df, on='account_id', how='outer')
df['loan'] = df['loan_id'].apply(lambda x: 'True' if pd.notna(x) else 'False')

# LOAN_AMOUNT
df.rename(columns={'amount':'loan_amount'}, inplace=True)

# LOAN PAYMENT
df.rename(columns={'payments':'loan_payment'}, inplace=True)

# LOAN_TERM
df.rename(columns={'terms':'loan_term'}, inplace=True)

# LOAN_STATUS
df['loan_status'] = df['status'].apply(lambda x: np.nan if pd.isna(x) else ("expired" if x == 'A' or x== 'B' else 'current')) # x: A, B, C, D, NaN / every row in loan_status

# LOAN_DEFAULT:
df['loan_default'] = df['status'].apply(lambda x: np.nan if pd.isna(x) else ('True' if x == 'B' or x== 'D' else 'False')) # x: A, B, C, D, NaN / every row in loan_status

# MAX_WITHDRAWAL
trans_df = pd.read_csv("../data/transactions.csv")

# return the maximum value of each account

# maximum value of debit
temp = trans_df[trans_df['type'] == 'debit'] # filter out
temp.groupby('account_id')

max_with = temp.groupby('account_id')['amount'].max()
max_with = pd.DataFrame(max_with)

df = pd.merge(df, max_with, on='account_id', how='left')
df.rename(columns = {'amount':'max_withdrawal'}, inplace=True)

min_with = temp.groupby('account_id')['amount'].min()
min_with = pd.DataFrame(min_with)

df = pd.merge(df, min_with, on='account_id', how='left')
df.rename(columns = {'amount':'min_withdrawal'}, inplace=True)

## CC payments
temp = trans_df[(trans_df['type'] == 'debit') & (trans_df['method'] == 'credit card')] # filter out
temp = temp.groupby('account_id')['amount'].count()

temp = pd.DataFrame(temp)

df = pd.merge(df, temp, on='account_id', how='left')
df.rename(columns = {'amount':'cc_payments'}, inplace=True)

# MAX BALANCE
temp = trans_df.groupby('account_id')['balance'].max() # filter out

max_bal = pd.DataFrame(temp)

df = pd.merge(df, max_bal, on='account_id', how='left')
df.rename(columns = {'balance':'max_balance'}, inplace=True)

# MIN BALANCE
temp = trans_df.groupby('account_id')['balance'].min() # filter out

min_bal = pd.DataFrame(temp)

df = pd.merge(df, min_bal, on='account_id', how='left')
df.rename(columns = {'balance':'min_balance'}, inplace=True)

columns_order = ['account_id', 'district_name', 'open_date','statement_frequency', 'num_customers', 'credit_cards', 'loan', 'loan_amount', 'loan_payment', 'loan_term', 'loan_status', 'loan_default','max_withdrawal','min_withdrawal','cc_payments','max_balance','min_balance']
df = df[columns_order]

df.to_csv('analytics_py.csv', index = False, encoding='utf-8') # open csv preview

File: codes-and-outputs/analytical.R
# import library
library('tidyverse')
library('reshape2')
library('dplyr')

# READ THE DATA
df <- read.csv("../data/accounts.csv")

# RENAME THE ID COLUMN
df <- df %>% rename(account_id = id)
df <- df %>% rename(open_date = date)

# REMOVE SPACES FROM COLUMN NAMES
colnames(df) <- gsub(" ", "", colnames(df))

# COUNT THE NUMBER OF MISSING VALUES IN EACH COLUMN; none
missing_values_num <- colSums(is.na(df))

# TYPE CAST THE MATRIX TO DATA FRAME
df <- as.data.frame(df)

# TYPE CAST THE DATE COLUMN
df$open_date <- as.Date(df$open_date, format = "%m/%d/%y")

# LOAD DISTRCICT DATA, MERGE TO THE MAIN DATA AND CHANGE THE COLUMN NAME
district_df <- read.csv("./districts_r.csv")
df <- left_join(df, district_df, by = "district_id")

# CHANGE THE COLUMN FROM DISTRICT_ID TO DISTRICT_NAME
df$district_id <- df$name
df <- df %>% rename(district_name = district_id)
df <- df[,c("account_id", "district_name", "open_date", "statement_frequency")]

# LOAD THE LINKS DATA FOR NUM_CUSTOMERS COLUMN
link_df <- read.csv("../data/links.csv")
temp <- data.frame(table(link_df$account_id))
names(temp) <- c("account_id", "num_customers")
df <- merge(df, temp, by.x = "account_id", by.y = "account_id", all.x = TRUE)

# CREDIT CARDS
card_df <- read.csv("../data/cards.csv")
temp <- merge(link_df, card_df, by.x = "id", by.y = "link_id", all.x = TRUE)
temp <- aggregate(id.y ~ account_id, data = temp, FUN = length)
names(temp) <- c("account_id", "credit_cards")
df <- left_join(df, temp, by = "account_id")

# LOAN
loan_df <- read.csv("./loans_r.csv")
df <- merge(df, loan_df, by = "account_id", all.x = TRUE)
df <- df %>%
  mutate(loan = ifelse(!is.na(loan_id), 'True', 'False'))
df <- df[, !(names(df) %in% c("loan_id"))]

# LOAN_AMOUNT
df <- df %>% rename(loan_amount = amount)

# LOAN PAYMENT
df <- df %>% rename(loan_payment = payments)

# LOAN TERM
df <- df %>% rename(loan_term = terms)

# LOAN STATUS
df$loan_status <- ifelse(is.na(df$status), NA,
                         ifelse(df$status %in% c('A', 'B'), 'expired', 'current'))

# LOAN_DEFAULT:
df$loan_default <- ifelse(is.na(df$status), NA,
                          ifelse(df$status %in% c('B', 'D'), 'True', 'False'))

# MAX WITHDRAWAL
trans_df <- read.csv("../data/transactions.csv")

max_withdrawal <- aggregate(trans_df$amount[trans_df$type == 'debit'],
                            by=list(account_id=trans_df$account_id[trans_df$type == 'debit']),
                            FUN=max)

colnames(max_withdrawal) <- c("account_id", "max_withdrawal")
df <- merge(df, max_withdrawal, by="account_id", all.x=TRUE)

# MIN WITHDRAWAL
min_withdrawal <- aggregate(trans_df$amount[trans_df$type == 'debit'],
                            by=list(account_id=trans_df$account_id[trans_df$type == 'debit']),
                            FUN=min)

colnames(min_withdrawal) <- c("account_id", "min_withdrawal")
df <- merge(df, min_withdrawal, by="account_id", all.x=TRUE)

## CC payments
temp <- trans_df[trans_df$type == 'debit' & trans_df$method == 'credit card', ]
cc_payments <- aggregate(temp$amount, by=list(account_id=temp$account_id), FUN=length)
colnames(cc_payments) <- c("account_id", "cc_payments")

df <- merge(df, cc_payments, by="account_id", all.x=TRUE)

# MAX BALANCE
temp <- aggregate(balance ~ account_id, data = trans_df, FUN = max)
max_bal <- data.frame(account_id = temp$account_id, max_balance = temp$balance)
df <- merge(df, max_bal, by = 'account_id', all.x = TRUE)
names(df)[names(df) == 'balance'] <- 'max_balance'

# MIN BALANCE
temp <- aggregate(balance ~ account_id, data = trans_df, FUN = min)
min_bal <- data.frame(account_id = temp$account_id, min_balance = temp$balance)
df <- merge(df, min_bal, by = 'account_id', all.x = TRUE)
names(df)[names(df) == 'balance'] <- 'min_balance'

# LAST CLEANING
columns_order <- c('account_id', 'district_name', 'open_date', 'statement_frequency', 'num_customers', 'credit_cards', 'loan', 'loan_amount', 'loan_payment', 'loan_term', 'loan_status', 'loan_default', 'max_withdrawal', 'min_withdrawal', 'cc_payments', 'max_balance', 'min_balance')

df <- df[columns_order]

# WRITE CSV FILE
write.csv(df, file = "analytics_r.csv", row.names = FALSE, fileEncoding = "UTF-8")